SixTrack/tools/astuce/astuce++.cpp

```diff
@@ -0,0 +1,708 @@
+	size_t Next_cd = 0;
+	size_t Next_dk = 0;
+
+	while(CurrentPosition < InputFileBuffer.size())
+	{
+		if(CurrentPosition == 0)
+		{
+			CurrentPosition = InputFileBuffer.find("+dk", CurrentPosition);
+		}
+		else
+		{
+			CurrentPosition = InputFileBuffer.find("\n+dk", CurrentPosition);
+		}
+
+		if(CurrentPosition == std::string::npos)
+		{
+			break;
+		}
+
+		LineEnd = InputFileBuffer.find("\n", CurrentPosition+1);
+
+		size_t NameStart = InputFileBuffer.rfind(" ", LineEnd);
+		std::string BlockName = FixCase(InputFileBuffer.substr(NameStart+1,LineEnd-NameStart-1));
+
+		//This could end at either the next +cd or the next +dk
+		//Find where the next one of these takes place and read until there.
+		Next_cd = InputFileBuffer.find("\n+cd", LineEnd);
+		Next_dk = InputFileBuffer.find("\n+dk", LineEnd);
+		CurrentPosition = std::min(Next_cd, Next_dk);
+
+		std::string BlockEntry = InputFileBuffer.substr(LineEnd+1, CurrentPosition-LineEnd-1);
+
+		//Now split this +dk block into lines
+		std::list<LineStorage> Lines;
+		LineStorage Entry;
+
+		size_t StartPosition = 0;
+		size_t BlockPosition = 0;
+
+		while(BlockPosition < BlockEntry.size())
+		{
+			BlockPosition = BlockEntry.find("\n", StartPosition);
+			if(BlockPosition == std::string::npos)
+			{
+				BlockPosition = BlockEntry.size();
+			}
+
+			Entry.Text = BlockEntry.substr(StartPosition, BlockPosition-StartPosition);
+			Entry.Enabled = true;
+
+			Lines.push_back(Entry);
+			StartPosition = BlockPosition+1;
+		}
+		DeckStorage.insert(std::make_pair(BlockName,Lines));
+	}
+
+	if(OutputFlag)
+	{
+		std::map<std::string, std::list<LineStorage> >::const_iterator it = DeckStorage.begin();
+		while(it!=DeckStorage.end())
+		{
+			*LogStream << it->first << " ";
+			it++;
+		}
+		*LogStream << std::endl;
+		*LogStream << "Total \"decks\": " << DeckStorage.size() << "\n" << std::endl;
+	}
+}
+
+void Astuce::ExpandIF()
+{
+	//start with +cd
+	//Remember we can have nested if levels!
+
+	std::map<std::string, std::list<LineStorage> >::iterator CallStorage_itr = CallStorage.begin();
+	//This loops over each +cd block
+	while(CallStorage_itr != CallStorage.end())
+	{
+		DoExpandIF(CallStorage_itr);
+		CallStorage_itr++;
+	}
+
+	//Then +dk
+	std::map<std::string, std::list<LineStorage> >::iterator DeckStorage_itr = DeckStorage.begin();
+	//This loops over each +dk block
+	while(DeckStorage_itr != DeckStorage.end())
+	{
+		DoExpandIF(DeckStorage_itr);
+		DeckStorage_itr++;
+	}
+}
+
+void Astuce::DoExpandIF(std::map<std::string, std::list<LineStorage> >::iterator Storage_itr)
+{
+	std::string BlockName = Storage_itr->first;
+
+	if(OutputFlag)
+	{
+		*LogStream << "Processing: " << BlockName << " with " << Storage_itr->second.size() << " lines." << std::endl;
+	}
+
+	EnabledStates.push(true);
+
+	std::list<LineStorage>::iterator Lines_itr = Storage_itr->second.begin();
+	while(Lines_itr != Storage_itr->second.end())
+	{
+		ProcessIfBlock(Lines_itr);
+		Lines_itr++;
+	}
+
+	EnabledStates.pop();
```

Comment:
If everything is OK, this *should* leave the stack exactly empty right?

It might be a good idea to check that this is the case, in case of missing `+if`. I guess if you have extra `+if`, the pop will fail and you'll have an exception, but one should also check the case that the stack is too long.